conveyor/adapter/base.go

```diff
@@ -17,80 +17,255 @@
+		} else {
+			itasks = swa.queue.RemoveAll()
+		}
+
+		log.Infof("[ StartProcessing ] Got %d new tasks", len(itasks))
+
+		if len(itasks) == 0 {
+			log.Info("[ StartProcessing ] No tasks. Sleep a little bit")
+			// TODO: do pretty wait
+			time.Sleep(50 * time.Millisecond)
+			continue
+		}
+
+		for _, itask := range itasks {
+			task, ok := itask.GetData().(queueTask)
+			if !ok {
+				panic(fmt.Sprintf("[ StartProcessing ] How does it happen? Wrong Type: %T", itask.GetData()))
+			}
+
+			if swa.processor == nil {
+				panic(fmt.Sprintf("[ StartProcessing ] Processor function wasn't provided"))
+			}
+
+			go swa.returnResponse(task)
+		}
+	}
+
+	swa.processingStopped <- true
 }
 
-// GetRespPayload implements Response method
-func (ar *AdapterResponse) GetRespPayload() interface{} {
-	return ar.respPayload
+var reqID uint64
+
+func atomicLoadAndIncrementUint64(addr *uint64) uint64 {
+	for {
+		val := atomic.LoadUint64(addr)
+		if atomic.CompareAndSwapUint64(addr, val, val+1) {
+			return val
+		}
+	}
+}
+
+// PushTask implements PulseConveyorAdapterTaskSink
+func (swa *CancellableQueueAdapter) PushTask(respSink AdaptorToSlotResponseSink,
+	elementID idType,
+	handlerID idType,
+	taskPayload interface{}) error {
+
+	cancelInfo := newCancelInfo(atomicLoadAndIncrementUint64(&reqID))
+	swa.taskHolder.add(cancelInfo, respSink.GetPulseNumber())
+
+	return swa.queue.SinkPush(
+		queueTask{
+			cancelInfo: cancelInfo,
+			task: AdapterTask{
+				respSink:    respSink,
+				elementID:   elementID,
+				handlerID:   handlerID,
+				taskPayload: taskPayload,
+			},
+		},
+	)
+}
+
+// CancelElementTasks: now cancels all pulseNumber's tasks
+func (swa *CancellableQueueAdapter) CancelElementTasks(pulseNumber idType, elementID idType) {
+	swa.taskHolder.stop(pulseNumber, false)
 }
 
-// AdapterNestedEvent contains info with adapter nested event
-type AdapterNestedEvent struct {
-	adapterID       idType
-	parentElementID idType
-	handlerID       idType
-	eventPayload    interface{}
+// CancelPulseTasks: now cancels all pulseNumber's tasks
+func (swa *CancellableQueueAdapter) CancelPulseTasks(pulseNumber idType) {
+	swa.taskHolder.stop(pulseNumber, false)
+}
+
+// FlushPulseTasks: now flush all pulseNumber's tasks
+func (swa *CancellableQueueAdapter) FlushPulseTasks(pulseNumber uint32) {
+	swa.taskHolder.stop(pulseNumber, true)
+}
+
+// FlushNodeTasks: now flush all tasks
+func (swa *CancellableQueueAdapter) FlushNodeTasks(nodeID idType) {
+	swa.taskHolder.stopAll(true)
+}
+
+// PushTask implements PulseConveyorAdapterTaskSink
+func (swa *CancellableQueueAdapter) returnResponse(cancellableTask queueTask) {
+	adapterTask := cancellableTask.task
+	event := swa.processor.Process(swa.adapterID, adapterTask, cancellableTask.cancelInfo)
+	if event.Flushed {
+		return
+	}
+	respSink := adapterTask.respSink
+	for nestedEvent := range event.NestedEventPayload {
+		respSink.PushNestedEvent(swa.adapterID, adapterTask.elementID, adapterTask.handlerID, nestedEvent)
```

Comment:
could we have NestedEvents end Response together?

Reply:
I mean, if we have Nested events should we send response?

Reply:
Yes, we should always send response (with one exception - if task was flush)